crates/sui-storage/src/indexes.rs

```diff
@@ -238,21 +257,85 @@ impl IndexStore {
         // Owner index
         batch.delete_batch(
             &self.tables.owner_index,
-            object_index_changes.deleted_owners.into_iter(),
+            object_index_changes.deleted_owners.iter().cloned(),
         )?;
         batch.delete_batch(
             &self.tables.dynamic_field_index,
             object_index_changes.deleted_dynamic_fields.into_iter(),
         )?;
         batch.insert_batch(
             &self.tables.owner_index,
-            object_index_changes.new_owners.into_iter(),
+            object_index_changes.new_owners.iter().cloned(),
         )?;
         batch.insert_batch(
             &self.tables.dynamic_field_index,
             object_index_changes.new_dynamic_fields.into_iter(),
         )?;
 
+        // Index coin info
+        let (input_coins, written_coins) = tx_coins;
+
+        // 1. Delete old owner if the object is deleted or transferred to a new owner,
+        // by looking at `object_index_changes.deleted_owners`.
+        // Objects in `deleted_owners` must be owned by `Owner::Address` befoer the tx,
+        // hence must appear in the tx inputs.
+        // They also mut be coin type (see `AuthorityState::commit_certificate`).
+        let coin_delete_keys = object_index_changes
+            .deleted_owners
+            .iter()
+            .filter_map(|(owner, obj_id)| {
+                // If it's not in `input_coins`, then it's not a coin type. Skip.
+                let coin_type_tag = input_coins.get(obj_id)?
+                .coin_type_maybe().unwrap_or_else(|| {
+                    panic!(
+                        "object_id: {:?} in input_coins is not a coin type, input_coins: {:?}, tx_digest: {:?}",
+                        obj_id, input_coins, digest
+                    )
+                });
```

Comment:
@longbowlu I may be naive in my thinking but it seems like if we extended  `owner_index` to store balance, we could've just deleted the key like how it is working today and don't need to do this.